### PR TITLE
[TECHDEBT] [P2P] P2P integration with utility for addressBook lookups - (Issue #270)

### LINKED ISSUE #270: [TECHDEBT] [P2P] P2P integration with utility for addressBook lookups
## Objective

Tend for the TODO

```go
// CLEANUP(drewsky): These functions will turn into more of a "ActorToAddrBook" when we have a closer
// integration with utility.
```

## Origin Document

**Techdebt** week emphasizes code health. Particularly, removing all `placeholder` and `hacky` code. 

The past few months, `shared/modules` have built up quite a few TODOs. 

The PR should remove all `non-blocked` TODOs

## Goals

- [ ] Address tech debt

## Deliverable

- [ ] Rename ValidatorMapToNetworkPeer to ActorMapToNetworkPeer or ActorsToNetworkPeer
- [ ] Look into the changing the interface call from ConsensusModule.ValidatorMap() to PersistenceModule.GetActorsAtHeight() because all the staked protocol level actors can help in broadcasting messages.

## Non-goals / Non-deliverables

- 

## General issue deliverables

- [ ] Update the appropriate CHANGELOG
- [ ] Update any relevant READMEs (local and/or global)
- [ ] Update any relevant global documentation & references
- [ ] If applicable, update the source code tree explanation
- [ ] If applicable, add or update a state, sequence or flowchart diagram using [mermaid](https://mermaid-js.github.io/mermaid/)

## [Optional] Testing Methodology

- \__REPLACE_ME_: Make sure to update the testing methodology appropriately\_
- **Task specific tests**: `make ...`
- **All tests**: `make test_all`
- **LocalNet**: verify a `LocalNet` is still functioning correctly by following the instructions at [docs/development/README.md](https://github.com/pokt-network/pocket/tree/main/docs/development)

---

**Creator**: @deblasis
**Co-Owners**: @Olshansk

### PR DESCRIPTION
**Pull request by deblasis**
## Description

This PR builds on top of #374 (the base branch is not `main` in order to highlight the delta of the changes) in order to tend to the TECHDEBT TODOs listed in #270 

## Issue

Fixes #270 

## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other 

## List of changes

- Consensus
  - `LeaderElectionModule`'s `electNextLeaderDeterministicRoundRobin` now uses `Persistence` to access the list of validators instead of the static `ValidatorMap`.
  - Updated tests (mock configs) accounting for the updated logic
- P2P
  - `ValidatorMapToAddrBook` renamed to `ActorToAddrBook`
  - `ValidatorToNetworkPeer` renamed to `ActorToNetworkPeer`

## Testing

- [x] `make develop_test`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [x] I have updated the corresponding CHANGELOG

### If Applicable Checklist
- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: consensus/consensus_tests/utils_test.go [hunk 1]
```diff
@@ -304,17 +304,22 @@ func basePersistenceMock(t *testing.T, _ modules.EventsChannel) *modulesMock.Moc
 	ctrl := gomock.NewController(t)
 	persistenceMock := modulesMock.NewMockPersistenceModule(ctrl)
 	persistenceContextMock := modulesMock.NewMockPersistenceRWContext(ctrl)
+	persistenceReadContextMock := modulesMock.NewMockPersistenceReadContext(ctrl)
 
 	persistenceMock.EXPECT().Start().Return(nil).AnyTimes()
 	persistenceMock.EXPECT().SetBus(gomock.Any()).Return().AnyTimes()
-	persistenceMock.EXPECT().NewReadContext(int64(-1)).Return(persistenceContextMock, nil).AnyTimes()
+	persistenceMock.EXPECT().NewReadContext(gomock.Any()).Return(persistenceReadContextMock, nil).AnyTimes()
```
Comment (deblasis) [inline consensus/consensus_tests/utils_test.go:311]: https://github.com/pokt-network/pocket/pull/374#discussion_r1041605270

File: consensus/leader_election/module.go [hunk 2]
```diff
@@ -60,5 +60,14 @@ func (m *leaderElectionModule) ElectNextLeader(message *typesCons.HotstuffMessag
 
 func (m *leaderElectionModule) electNextLeaderDeterministicRoundRobin(message *typesCons.HotstuffMessage) typesCons.NodeId {
```
Comment (deblasis) [inline consensus/leader_election/module.go:61]: https://github.com/pokt-network/pocket/pull/374#discussion_r1041612218

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -55,6 +55,10 @@ jobs:
         run: |
           set -euo pipefail
           go test -p 1 -json ./... -covermode=count -coverprofile=coverage.out 2>&1 | tee test_results.json
+      - name: Sanitize test results
```
[the remark this answers is not in the document]
Comment (deblasis) [reply .github/workflows/main.yml]: Done

File: consensus/leader_election/module.go (hunk 2 above)
Comment (deblasis) [reply consensus/leader_election/module.go]: https://github.com/pokt-network/pocket/pull/378/commits/dd3020e9343972561ad00cee123e65f86357b97a

File: consensus/consensus_tests/utils_test.go (hunk 1 above)
Comment (deblasis) [reply consensus/consensus_tests/utils_test.go]: This has been addressed. Now we are using the context from `Utility`

https://github.com/pokt-network/pocket/pull/378/commits/8f837201f2f2ec9132e72993bd98b1219cc2d88a

Comment (deblasis) [reply consensus/consensus_tests/utils_test.go]: I have a couple of observations I'd like to have your 2cents about: 
- is it by design that Utility always returns a RWContext and not a ReadOnly one?
- I find it weird that we instantiate the context passing in the height and then we have to specify the height again for all the calls like:

```go
	uCtx, err := m.GetBus().GetUtilityModule().NewContext(int64(message.Height))
	if err != nil {
		return typesCons.NodeId(0), err
	}
	vals, err := uCtx.GetPersistenceContext().GetAllValidators(int64(message.Height))
	if err != nil {
		return typesCons.NodeId(0), err
	}
```

File: consensus/consensus_tests/utils_test.go [hunk 4]
```diff
@@ -306,17 +306,22 @@ func basePersistenceMock(t *testing.T, _ modules.EventsChannel) *modulesMock.Moc
...
 	persistenceMock.EXPECT().ReleaseWriteContext().Return(nil).AnyTimes()
 
 	// The persistence context should usually be accessed via the utility module within the context
 	// of the consensus module. This one is only used when loading the initial consensus module
 	// state; hence the `-1` expectation in the call above.
 	persistenceContextMock.EXPECT().Close().Return(nil).AnyTimes()
-	persistenceContextMock.EXPECT().GetLatestBlockHeight().Return(uint64(0), nil).AnyTimes()
+	persistenceReadContextMock.EXPECT().GetLatestBlockHeight().Return(uint64(0), nil).AnyTimes()
+
+	// here we are mocking 4 actors since we are considering the 4 nodes in the LocalNet genesis
```
Comment (Olshansk) [inline consensus/consensus_tests/utils_test.go:322]: I know it might be overkill, but if we use a repeat function (not available for structs by default), you can reuse the `numValidators` parameter: https://go.dev/play/p/uWDXFeh5d_o because that's what runTimeMgr is dependant on

File: consensus/doc/CHANGELOG.md [hunk 5]
```diff
@@ -7,10 +7,14 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
-## [0.0.0.12] - 2022-12-06
+## [0.0.0.13] - 2022-12-06
```
Comment (Olshansk) [inline consensus/doc/CHANGELOG.md:10]: This has already been merged in. Reorder?

File: p2p/CHANGELOG.md [hunk 6]
```diff
@@ -6,11 +6,15 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
-
-## [0.0.0.10] 2022-12-08
+## [0.0.0.11] 2022-12-08
```
Comment (Olshansk) [inline p2p/CHANGELOG.md:9]:
```suggestion

## [0.0.0.11] 2022-12-08
```

File: p2p/utils.go [hunk 7]
```diff
@@ -9,12 +9,10 @@ import (
...
 )
 
-// CLEANUP(drewsky): These functions will turn into more of a "ActorToAddrBook" when we have a closer
-// integration with utility.
-func ValidatorMapToAddrBook(cfg modules.P2PConfig, validators map[string]modules.Actor) (typesP2P.AddrBook, error) {
+func ActorToAddrBook(cfg modules.P2PConfig, actors map[string]modules.Actor) (typesP2P.AddrBook, error) {
 	book := make(typesP2P.AddrBook, 0)
-	for _, v := range validators {
-		networkPeer, err := ValidatorToNetworkPeer(cfg, v)
+	for _, v := range actors {
```
Comment (Olshansk) [inline p2p/utils.go:14]: s/v/a

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: consensus/consensus_tests/utils_test.go (hunk 1 above)
Comment (Olshansk) [reply consensus/consensus_tests/utils_test.go]: Great questions! tl;dr Legacy. Adding @gokutheengineer for context.

> is it by design that Utility always returns a RWContext and not a ReadOnly one?

No, I support adding the ability to instantiate a utility context with a read context.

> I find it weird that we instantiate the context passing in the height and then we have to specify the height again for all the calls like:

Agreed. I would say this:
1. A read context should be able to get instantiated without a height -> it only lets you read committed data on disk at any height.
2. A write context needs to be able to get instantiated with a height -> writes/update will happen, by default, at that height but not committed until the context is committed. Reads can happen for any height. Readds for he height provided for writing reads from the "context" (not committed.

@deblasis Can you create a ticket for this?

Comment (deblasis) [reply consensus/consensus_tests/utils_test.go]: > Agreed. I would say this:
> 
> 1. A read context should be able to get instantiated without a height -> it only lets you read committed data on disk at any height.
> 2. A write context needs to be able to get instantiated with a height -> writes/update will happen, by default, at that height but not committed until the context is committed. Reads can happen for any height. Readds for he height provided for writing reads from the "context" (not committed.

🤔 

## R/W

As far as I understand, a `RWContext` is supposed to be working in an append-only way. The state at a given `height` is immutable after it is committed / the block is finalized.
That said, it looks like the `RWContext` should not get the `height` argument in the constructor and also in the "setter" methods.
Why? 
Because we are writing a state where `height` = `currentheight`. 
Do we ever want to be able to overwrite a previous state? I don't think so but correct me if I am wrong.

## ReadOnly

For read only getters, it makes sense to have a height because technically you could be asking for a "view" of the state at any `height` since genesis.
The way I see the interface is that you instantiate providing a `height` and then all the getters from that context would be at that height. 

Let me know your thoughts on this.

File: p2p/utils.go (hunk 7 above)
Comment (deblasis) [reply p2p/utils.go]: https://github.com/pokt-network/pocket/pull/378/commits/f70466a8a3844732bd8bcf9b700953af88263126

File: consensus/consensus_tests/utils_test.go (hunk 4 above)
Comment (deblasis) [reply consensus/consensus_tests/utils_test.go:322]: Yeah, me likey generics but I'd rather go with something like this https://github.com/pokt-network/pocket/pull/378/commits/7c051965290273863892406755f36914d7c83132

File: consensus/leader_election/module.go [hunk 8]
```diff
@@ -55,10 +55,23 @@ func (m *leaderElectionModule) GetBus() modules.Bus {
...
-func (m *leaderElectionModule) electNextLeaderDeterministicRoundRobin(message *typesCons.HotstuffMessage) typesCons.NodeId {
+func (m *leaderElectionModule) electNextLeaderDeterministicRoundRobin(message *typesCons.HotstuffMessage) (typesCons.NodeId, error) {
 	value := int64(message.Height) + int64(message.Round) + int64(message.Step) - 1
-	return typesCons.NodeId(value%int64(len(m.GetBus().GetConsensusModule().ValidatorMap())) + 1)
+
+	uCtx, err := m.GetBus().GetUtilityModule().NewContext(int64(message.Height))
+	if err != nil {
+		return typesCons.NodeId(0), err
+	}
+	vals, err := uCtx.GetPersistenceContext().GetAllValidators(int64(message.Height))
```
Comment (Olshansk) [inline consensus/leader_election/module.go:72]: Let's create a read context `m.GetBus().GetPersistenceModule().NewReadContext(int64(height))` like you did in #374. Keeps it simpler.

File: p2p/module.go [hunk 9]
```diff
@@ -129,7 +129,7 @@ func (m *p2pModule) Start() error {
 	if m.GetBus().GetPersistenceModule() == nil {
 		// we are getting called by the client and we use the "legacy behaviour"
 		// TODO (team): improve this.
-		addrBook, err = ValidatorMapToAddrBook(m.p2pCfg, m.bus.GetConsensusModule().ValidatorMap())
+		addrBook, err = ActorToAddrBook(m.p2pCfg, m.bus.GetConsensusModule().ValidatorMap())
```
Comment (Olshansk) [inline p2p/module.go:132]: GetBus()

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: consensus/consensus_tests/utils_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (deblasis) [reply consensus/consensus_tests/utils_test.go]: removed altogether. Hopefully the code is clear enough now that we are calling `makeMockActors(numValidators)`

File: consensus/leader_election/module.go (hunk 8 above)
Comment (deblasis) [reply consensus/leader_election/module.go:72]: We are doing it this way because of this
image

remember we discussed it and you asked me to get the context from Utility?

**Comment by deblasis** [comment]
I think this is ready for another lookie.

File: consensus/consensus_tests/utils_test.go [hunk 10]
```diff
@@ -306,17 +306,21 @@ func basePersistenceMock(t *testing.T, _ modules.EventsChannel) *modulesMock.Moc
...
+	persistenceMock.EXPECT().NewReadContext(int64(-1)).Return(persistenceReadContextMock, nil).AnyTimes()
 	persistenceMock.EXPECT().ReleaseWriteContext().Return(nil).AnyTimes()
 
 	// The persistence context should usually be accessed via the utility module within the context
 	// of the consensus module. This one is only used when loading the initial consensus module
 	// state; hence the `-1` expectation in the call above.
 	persistenceContextMock.EXPECT().Close().Return(nil).AnyTimes()
-	persistenceContextMock.EXPECT().GetLatestBlockHeight().Return(uint64(0), nil).AnyTimes()
+	persistenceReadContextMock.EXPECT().GetLatestBlockHeight().Return(uint64(0), nil).AnyTimes()
+
```
Comment (Olshansk) [inline consensus/consensus_tests/utils_test.go:321]:
```suggestion
```

File: consensus/consensus_tests/utils_test.go (hunk 1 above)
Comment (Olshansk) [reply consensus/consensus_tests/utils_test.go]: > Let me know your thoughts on this.

I think that you are completely, 💯 %, whole heartedly w̶r̶o̶n̶g̶ correct

tl;dr
- Write interface should accept a height
- Read interface should not accept a height
- Getters should accept a height
- Setters should not accept a height

🤝 ?

Comment (deblasis) [reply consensus/consensus_tests/utils_test.go]: 😂 LOL

Ok, I just want to make sure I understand. I have two additional stupid questions:

>- Write interface should accept a height

1. If we can have only one RW context and we are working on the current block up until we commit it, why do we need a height? 

>- Read interface should not accept a height
>- Getters should accept a height

2. If the Read interface is meant to access the state at a certain height, why do we want to have to specify the height at every getter call and not once per-readcontext when we instantiate it?

>- Setters should not accept a height

👌

With these answers I guess I'll have enough context to write a mini-spec for the new ticket

**Comment by Olshansk** [comment]
@deblasis Can you create a ticket for https://github.com/pokt-network/pocket/pull/378#discussion_r1047894172 for you to handle?

**Comment by deblasis** [comment]
@Olshansk 

> @deblasis Can you create a ticket for [#378 (comment)](https://github.com/pokt-network/pocket/pull/378#discussion_r1047894172) for you to handle?

https://github.com/pokt-network/pocket/pull/378#discussion_r1049055338

**Comment by deblasis** [comment]
> @deblasis Can you create a ticket for [#378 (comment)](https://github.com/pokt-network/pocket/pull/378#discussion_r1047894172) for you to handle?

This is now tracked in #406
